File: tree/508.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


func findFrequentTreeSum(root *TreeNode) []int {
    if root == nil {
        return nil
    }
    var (
        index int
        ret []int
    )
    RetMap := make(map[int]int)
    findSum(root, &RetMap)
    fmt.Println(RetMap)
    for _, l := range RetMap {
        if l > index {
            index = l
        }
    }
    for i, l := range RetMap {
        if index == l {
            ret = append(ret, i)
        }
    }
    return ret
}

func findSum(root *TreeNode, retMap *map[int]int) int{
    ret := root.Val
    if root.Left != nil {
        //自底向上累加
        ret += findSum(root.Left, retMap)
    }
    
    if root.Right != nil {
        ret += findSum(root.Right, retMap)
    }
    //加入结果集
    if _, ok := (*retMap)[ret]; !ok {
        (*retMap)[ret] = 1
    }else {
        (*retMap)[ret]++
    }
    //返回该节点的结果
    return ret
}

File: tree/508.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> findFrequentTreeSum(TreeNode* root) {
        if (!root)
            return ret_;
        recur(root);
        int count = 0, ret = 0;
        for (auto iter = m_.begin(); iter != m_.end(); ++iter) {
            count = std::max(count, iter->second);
        }
        for (auto iter = m_.begin(); iter != m_.end(); ++iter) {
            if (iter->second == count) {
                ret_.emplace_back(iter->first);
            }
        }
        return ret_;
    }
private:
    std::unordered_map<int, int> m_;
    std::vector<int> ret_;
    int recur(TreeNode* root) {
        if (!root)
            return 0;
        int left = recur(root->left);
        int right = recur(root->right);
        int sum = root->val + left + right;
        ++m_[sum];
        return sum;
    }
};
